Sources/Amplitude/Utilities/EventPipeline.swift

```diff
@@ -54,32 +54,35 @@ public class EventPipeline {
         eventCount = 0
         guard let storage = self.storage else { return }
         storage.rollover()
-        guard let eventFiles: [URL]? = storage.read(key: StorageKey.EVENTS) else { return }
-        cleanupUploads()
-        if pendingUploads == 0 {
-            for eventFile in eventFiles! {
-                guard let eventsString = storage.getEventsString(eventBlock: eventFile) else {
-                    continue
-                }
-                if eventsString.isEmpty {
-                    continue
-                }
-                let uploadTask = httpClient.upload(events: eventsString) { [weak self] result in
-                    let responseHandler = storage.getResponseHandler(
-                        configuration: self!.amplitude.configuration,
-                        eventPipeline: self!,
-                        eventBlock: eventFile,
-                        eventsString: eventsString
-                    )
-                    responseHandler.handle(result: result)
-                    self?.cleanupUploads()
-                }
-                if let upload = uploadTask {
-                    add(uploadTask: UploadTaskInfo(events: eventsString, task: upload))
+        guard let eventFiles: [URL] = storage.read(key: StorageKey.EVENTS) else { return }
+        for eventFile in eventFiles {
+            guard existingUploadTaskInfo(for: eventFile) == nil else {
+                continue
+            }
+            guard let eventsString = storage.getEventsString(eventBlock: eventFile) else {
+                continue
+            }
+            if eventsString.isEmpty {
+                continue
+            }
+            let uploadTask = httpClient.upload(events: eventsString) { [weak self] result in
+                guard let self else {
+                    return
                 }
+                let responseHandler = storage.getResponseHandler(
+                    configuration: self.amplitude.configuration,
+                    eventPipeline: self,
+                    eventBlock: eventFile,
+                    eventsString: eventsString
+                )
+                responseHandler.handle(result: result)
+                self.completeUpload(for: eventFile)
+            }
+            if let upload = uploadTask {
+                add(uploadTask: UploadTaskInfo(events: eventsString, task: upload), for: eventFile)
             }
-            completion?()
         }
+        completion?()
```

Comment:
Does this still run into a case where the completion has not been called yet and a new flush gets triggered at the same time? It'll still re-upload all files (although the possibility of this happening is super low)

Reply:
In such cases, does it make sense to skip new flushes until the completion of an existing flush is not completed?

Reply:
This completion handler is only called from our tests, I don't think EventPipeline is externally accessible. It is possible for multiple flushes to occur simultaneously, though this occurred in the previous code as well, albeit in fewer circumstances (when all http requests are complete vs at any time). 

I think this behavior is consistent with the expectation that a flush completes when all events prior to the flush have begun upload, which this still guarantees. We should never repeat work as the task control is handled in a single queue, so a given file will only be uploaded once no matter which flush started it. Can you be more specific as to how this would re-upload all files?

Reply:
Actually, I guess there is a slight possibility of duplicate uploads if two flushes are run simultaneously and both hit `existingUploadTaskInfo` before actually adding the task. We can just run the whole `for` loop body on the upload queue to prevent this.

Reply:
I've made the changes to make the for loop body run on the uploadQueue, which should resolve any duplicate upload issues. Please let me know if I'm missing anything!

Reply:
ah nvm, i thought the file deletion happened after all files were flushed. Just check that it happens inside the response handler.